### PR TITLE
[flutter_tools] update build rules to depend on subset of package_config contents

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Split from https://github.com/flutter/flutter/pull/66776

Even if pub does not change the packge_config contents, it will still update a timestamp in one of the fields. This causes unnecessary rebuilds. To fix this, generate an additional file when running pub get that only contains the relevant fields and then update the KernelSnapshot rule to depend on it only.

### CODE REVIEW

File: packages/flutter_tools/test/integration.shard/test_test.dart [hunk 1]
```diff
@@ -5,79 +5,84 @@
...
 
-import 'package:flutter_tools/src/artifacts.dart';
 import 'package:flutter_tools/src/base/file_system.dart';
 import 'package:flutter_tools/src/base/io.dart';
-import 'package:flutter_tools/src/cache.dart';
-import 'package:flutter_tools/src/globals.dart' as globals;
 
-import '../../src/common.dart';
-import '../../src/context.dart';
+import '../src/common.dart';
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/test_test.dart:11]: For the second part of this change this test needs to be changed to be more selective about running pub

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 2]
```diff
@@ -480,4 +499,18 @@ class _DefaultPub implements Pub {
...
+
+  // Subset the package config file to only the parts that are relevant for
+  // rerunning the dart compiler.
+  String _computePackageConfigSubset(PackageConfig packageConfig, FileSystem fileSystem) {
+    final StringBuffer buffer = StringBuffer();
+    for (final Package package in packageConfig.packages) {
+      buffer.write(package.name);
+      buffer.write(package.languageVersion);
+      buffer.write(package.root);
+      buffer.write(package.packageUriRoot);
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/dart/pub.dart:511]: Should these be separated by newlines?  Doesn't matter now, but it'll be hard to parse back as a single line if we ever need to.

File: packages/flutter_tools/lib/src/dart/pub.dart [hunk 3]
```diff
@@ -458,13 +459,31 @@ class _DefaultPub implements Pub {
     return environment;
   }
 
-  /// Insert the flutter_gen synthetic package into the package configuration file if
-  /// there is an l10n.yaml.
-  Future<void> _updatePackageConfig(File packageConfigFile, Directory generatedDirectory) async {
-    if (!packageConfigFile.existsSync()) {
+
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/dart/pub.dart:462]: Extra newline

**Review by jmagman — APPROVED** [review]
LGTM I think?

File: packages/flutter_tools/lib/src/dart/pub.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/dart/pub.dart]: Done

File: packages/flutter_tools/lib/src/dart/pub.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/dart/pub.dart]: Done

**Comment by jonahwilliams** [comment]
Added some more clarification on "synthetic" package in pub.dart
